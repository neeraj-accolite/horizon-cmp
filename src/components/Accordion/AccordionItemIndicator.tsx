import React from 'react';
import Image from 'next/image';

const downArrow = '/downArrow.svg';
const upArrow = '/upArrow.svg';

interface AccordionItemIndicatorProps {
  isOpen: boolean;
  isDisabled?: boolean;
}

const AccordionItemIndicator: React.FC<AccordionItemIndicatorProps> = ({
  isOpen,
}) => {
  return (
    <div
      className={`ease transition-transform duration-300 ${
        isOpen ? 'rotate-90' : 'rotate-0'
      }`}
    >
      <Image
        src={isOpen ? upArrow : downArrow}
        alt="Arrow Icon"
        width={16}
        height={16}
      />
    </div>
  );
};

export default AccordionItemIndicator;
