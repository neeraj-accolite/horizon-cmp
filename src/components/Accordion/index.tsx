export { AccordionComp } from './Accordion';
export type { AccordionProps } from './Accordion.model';
