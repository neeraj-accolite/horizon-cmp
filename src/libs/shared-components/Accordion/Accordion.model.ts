import { AccordionProps as NextUIAccordionProps } from '@nextui-org/react';
import { JSXElementConstructor, ReactElement } from 'react';

export interface AccordionItem {
  title: string;
  children?: ReactElement<unknown, string | JSXElementConstructor<unknown>>;
}

export interface AccordionProps extends NextUIAccordionProps {
  items: AccordionItem[];
  height?: string;
  width?: string;
  multiExpand?: boolean; // To allow multiple items to be expanded at the same t  onToggle?: (index: number) => void; // Callback function when an item is toggled
  className?: string; // Custom class name for styling
  style?: React.CSSProperties; // Inline styles for the accordion
  showSeparators?: boolean; // To show or hide separators between item
  dividerColor?: string;
}
