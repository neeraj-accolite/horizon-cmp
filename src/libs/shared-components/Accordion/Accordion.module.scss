.divider {
    @apply border border-solid h-px border-accordion-divider-color;
  }