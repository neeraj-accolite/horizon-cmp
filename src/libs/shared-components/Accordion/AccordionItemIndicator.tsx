import React from 'react';
import ArrowLeft from '../Icons/IconList/ArrowLeft';
import DownArrow from '../Icons/IconList/DownArrow';

interface AccordionItemIndicatorProps {
  isOpen: boolean;
}

const AccordionItemIndicator: React.FC<AccordionItemIndicatorProps> = ({
  isOpen,
}) => {
  return (
    <div
      className={`ease pb-2 pt-4 transition-transform duration-300 ${
        isOpen ? 'rotate-90' : 'rotate-0'
      }`}
    >
      {isOpen ? (
        <DownArrow size={16} data-testid="down-arrow-icon" />
      ) : (
        <ArrowLeft size={16} data-testid="arrow-left-icon" />
      )}
    </div>
  );
};

export default AccordionItemIndicator;
