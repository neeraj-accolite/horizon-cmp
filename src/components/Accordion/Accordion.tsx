import React from 'react';
import {
  Accordion,
  AccordionItem,
  AccordionItemIndicatorProps,
} from '@nextui-org/accordion';
import { AccordionProps } from './Accordion.model';
import AccordionItemIndicator from './AccordionItemIndicator';

export const AccordionComp: React.FC<AccordionProps> = ({
  items,
  className,
  dividerColor = '#8E8C99',
  showSeparators = false,
}) => {
  return (
    <div className={`${className} rounded-lg p-4 shadow-md`}>
      <Accordion
        aria-label="Accordion"
        showDivider={showSeparators}
        dividerProps={{ className: `bg-[${dividerColor}] h-0.5` }}
      >
        {items.map((item, index) => (
          <AccordionItem
            key={index}
            aria-label={item.title}
            title={
              <span className="text-base font-bold leading-[24px]">
                {item.title}
              </span>
            }
            indicator={(props: AccordionItemIndicatorProps) => {
              return <AccordionItemIndicator isOpen={props.isOpen || false} />;
            }}
          >
            <div className="mb-3 text-[16px] font-normal leading-[24px]">
              {item.children && (
                <div className="mt-0 p-0">
                  <ul className="mt-0 pl-0">{item.children}</ul>
                </div>
              )}
            </div>
          </AccordionItem>
        ))}
      </Accordion>
    </div>
  );
};
