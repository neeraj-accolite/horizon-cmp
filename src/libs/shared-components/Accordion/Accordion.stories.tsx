import React from 'react';
import { Meta, StoryFn } from '@storybook/react';
import Accordion from './Accordion';
import { AccordionProps } from './Accordion.model';

export default {
  title: 'Components/Accordion',
  component: Accordion,
} as Meta;

const Template: StoryFn<AccordionProps> = (args) => <Accordion {...args} />;

export const SingleAccordion = Template.bind({});
SingleAccordion.args = {
  items: [
    {
      title: 'Single Accordion Item',
      children: (
        <ul>
          <li>Child Item 1</li>
          <li>Child Item 2</li>
        </ul>
      ),
    },
  ],
  height: 'auto',
  width: '100%',
  multiExpand: false,
  className: 'custom-accordion',
  showSeparators: true,
};

export const MultiNodeAccordion = Template.bind({});
MultiNodeAccordion.args = {
  items: [
    {
      title: 'Accordion Item 1',
      children: (
        <ul>
          <li>Child Item 1</li>
          <li>Child Item 2</li>
        </ul>
      ),
    },
    {
      title: 'Accordion Item 2',
      children: (
        <ul>
          <li>Child Item 1</li>
          <li>Child Item 2</li>
        </ul>
      ),
    },
    {
      title: 'Accordion Item 3',
      children: (
        <ul>
          <li>Child Item 1</li>
          <li>Child Item 2</li>
        </ul>
      ),
    },
  ],
  height: 'auto',
  width: '100%',
  showSeparators: true,
  dividerColor: '#8E8C99',
};
