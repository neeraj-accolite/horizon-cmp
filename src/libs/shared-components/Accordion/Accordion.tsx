import React from 'react';
import {
  Accordion as AccordionComp,
  AccordionItem,
  AccordionItemIndicatorProps,
} from '@nextui-org/accordion';
import { AccordionProps } from './Accordion.model';
import AccordionItemIndicator from './AccordionItemIndicator';
import { PlumbingVector } from '../Icons';
import styles from './Accordion.module.scss';

const Accordion: React.FC<AccordionProps> = ({
  items,
  className,
  vector,
  showSeparators = true,
}) => {
  return (
    <div className={`${className} rounded-lg px-4 py-8 shadow-md`}>
      <AccordionComp
        aria-label="Accordion"
        showDivider={showSeparators}
        dividerProps={{
          className: `${styles.divider}`,
        }}
      >
        {items.map((item, index) => (
          <AccordionItem
            key={index}
            aria-label={item.title}
            title={
              <div className="flex items-center pb-2 pt-4 text-lg">
                {vector && <PlumbingVector fill="currentColor" size={32} />}
                <h5 className="ml-1">{item.title}</h5>
              </div>
            }
            indicator={(props: AccordionItemIndicatorProps) => {
              return <AccordionItemIndicator isOpen={props.isOpen || false} />;
            }}
          >
            <div className="mb-3 text-base font-normal leading-6">
              {item.children && (
                <div className="mt-0 p-0">
                  <ul className="mt-0 pl-0">{item.children}</ul>
                </div>
              )}
            </div>
          </AccordionItem>
        ))}
      </AccordionComp>
    </div>
  );
};

export default Accordion;
