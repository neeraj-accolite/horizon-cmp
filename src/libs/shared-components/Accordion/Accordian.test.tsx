import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Accordion from './Accordion';
import { AccordionProps } from './Accordion.model';
import AccordionItemIndicator from './AccordionItemIndicator';

const items: AccordionProps['items'] = [
  {
    title: 'Accordion Item 1',
    children: (
      <ul>
        <li>Child Item 1</li>
        <li>Child Item 2</li>
      </ul>
    ),
  },
  {
    title: 'Accordion Item 2',
    children: (
      <ul>
        <li>Child Item 1</li>
        <li>Child Item 2</li>
      </ul>
    ),
  },
  {
    title: 'Accordion Item 3',
    children: (
      <ul>
        <li>Child Item 1</li>
        <li>Child Item 2</li>
      </ul>
    ),
  },
];
jest.mock('next/image', () => ({
  __esModule: true,
  default: (
    props: React.JSX.IntrinsicAttributes &
      React.ClassAttributes<HTMLImageElement> &
      React.ImgHTMLAttributes<HTMLImageElement>,
  ) => <img {...props} />,
}));

describe('MultiNodeAccordion', () => {
  test('matches snapshot', () => {
    const { asFragment } = render(
      <Accordion items={items}>
        <ul>
          <li>Child Item 1</li>
          <li>Child Item 2</li>
        </ul>
      </Accordion>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders all accordion items', () => {
    render(
      <Accordion items={items}>
        <ul>
          <li>Child Item 1</li>
          <li>Child Item 2</li>
        </ul>
      </Accordion>,
    );
    items.forEach((item) => {
      expect(screen.getByText(item.title)).toBeInTheDocument();
    });
  });
  it('should render down arrow when isOpen is false', () => {
    render(<AccordionItemIndicator isOpen={false} />);

    const image = screen.getByAltText('Arrow Icon');
    expect(image).toHaveAttribute('src', '/downArrow.svg');
  });

  it('should render up arrow when isOpen is true', () => {
    render(<AccordionItemIndicator isOpen={true} />);

    const image = screen.getByAltText('Arrow Icon');
    expect(image).toHaveAttribute('src', '/upArrow.svg');
  });
});
