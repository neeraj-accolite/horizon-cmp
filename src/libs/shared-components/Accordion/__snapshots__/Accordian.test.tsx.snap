// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiNodeAccordion matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="undefined rounded-lg p-4 shadow-md"
  >
    <div
      aria-label="Accordion"
      class="px-2 w-full"
      data-orientation="vertical"
    >
      <div
        aria-label="Accordion Item 1"
      >
        <h2>
          <button
            aria-expanded="false"
            class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
            id=":r0:"
            type="button"
          >
            <div
              class="flex-1 flex flex-col text-start"
            >
              <span
                class="text-foreground text-medium"
              >
                <span
                  class="text-base font-bold leading-6"
                >
                  Accordion Item 1
                </span>
              </span>
            </div>
            <span
              aria-hidden="true"
              class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
            >
              <div
                class="ease transition-transform duration-300 rotate-0"
              >
                <img
                  alt="Arrow Icon"
                  height="16"
                  src="/downArrow.svg"
                  width="16"
                />
              </div>
            </span>
          </button>
        </h2>
      </div>
      <div
        aria-label="Accordion Item 2"
      >
        <h2>
          <button
            aria-expanded="false"
            class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
            id=":r2:"
            type="button"
          >
            <div
              class="flex-1 flex flex-col text-start"
            >
              <span
                class="text-foreground text-medium"
              >
                <span
                  class="text-base font-bold leading-6"
                >
                  Accordion Item 2
                </span>
              </span>
            </div>
            <span
              aria-hidden="true"
              class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
            >
              <div
                class="ease transition-transform duration-300 rotate-0"
              >
                <img
                  alt="Arrow Icon"
                  height="16"
                  src="/downArrow.svg"
                  width="16"
                />
              </div>
            </span>
          </button>
        </h2>
      </div>
      <div
        aria-label="Accordion Item 3"
      >
        <h2>
          <button
            aria-expanded="false"
            class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
            id=":r4:"
            type="button"
          >
            <div
              class="flex-1 flex flex-col text-start"
            >
              <span
                class="text-foreground text-medium"
              >
                <span
                  class="text-base font-bold leading-6"
                >
                  Accordion Item 3
                </span>
              </span>
            </div>
            <span
              aria-hidden="true"
              class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
            >
              <div
                class="ease transition-transform duration-300 rotate-0"
              >
                <img
                  alt="Arrow Icon"
                  height="16"
                  src="/downArrow.svg"
                  width="16"
                />
              </div>
            </span>
          </button>
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
